pkg/component/runtime/manager.go

```diff
@@ -333,6 +341,69 @@ func (m *Manager) PerformAction(ctx context.Context, unit component.Unit, name s
 	return respBody, nil
 }
 
+// PerformDiagnostics executes the diagnostic action for the provided units. If no units are provided then
+// it performs diagnostics for all current units.
+func (m *Manager) PerformDiagnostics(ctx context.Context, units ...component.Unit) []ComponentUnitDiagnostic {
+	// build results from units
+	var results []ComponentUnitDiagnostic
+	if len(units) > 0 {
+		for _, u := range units {
+			r := m.getRuntimeFromUnit(u)
+			if r == nil {
+				results = append(results, ComponentUnitDiagnostic{
+					Unit: u,
+					Err:  ErrNoUnit,
+				})
+			} else {
+				results = append(results, ComponentUnitDiagnostic{
+					Component: r.currComp,
+					Unit:      u,
+				})
+			}
+		}
+	} else {
+		m.mx.RLock()
+		for _, r := range m.current {
+			for _, u := range r.currComp.Units {
+				var err error
+				if r.currComp.Err != nil {
+					err = r.currComp.Err
+				} else if u.Err != nil {
+					err = u.Err
+				}
+				if err != nil {
+					results = append(results, ComponentUnitDiagnostic{
+						Component: r.currComp,
+						Unit:      u,
+						Err:       err,
+					})
+				} else {
+					results = append(results, ComponentUnitDiagnostic{
+						Component: r.currComp,
+						Unit:      u,
+					})
+				}
+			}
+		}
+		m.mx.RUnlock()
+	}
+
+	for i, r := range results {
+		if r.Err != nil {
+			// already in error don't perform diagnostics
+			continue
+		}
+		diag, err := m.performDiagAction(ctx, r.Unit)
+		if err != nil {
+			r.Err = err
+		} else {
+			r.Results = diag
+		}
```

Comment:
[Suggestion]
It if you remove the if checks, the outcome would still be the same as `diag`is `nil` on error, `err` is nil when `diag` isn't.